lib/tasks/migrate.rake

```diff
@@ -1,76 +1,145 @@
+    # from logs
+    logs.each do |level, resource, message|
+      keywords_set["PuppetFailed:#{resource}"] = true if level == "err" && resource != "Puppet"
+    end
+    keywords_set.keys
   end
 
-  def summary(metrics)
+  def summary(metrics, status)
     change, failure, nochange = metrics_values(metrics)
     {
-      :foreman => {
-        :change => change,
-        :nochange => nochange,
-        :failure => failure,
+      foreman: {
+        change: change,
+        nochange: nochange,
+        failure: failure,
       },
-      :native => metrics[:resources],
+      native: metrics[:resources] || status || {},
     }
   end
 
   def metrics_values(metrics)
-    change = metrics.dig("events", "success")
-    failure = metrics.dig("events", "failure")
-    total = metrics.dig("events", "total")
+    return [0, 0, 0] if metrics.empty?
+    change = metrics.dig("events", "success") || 0
+    failure = metrics.dig("events", "failure") || 0
+    total = metrics.dig("events", "total") || 0
     nochange = total - change - failure
     [change, failure, nochange]
+  rescue StandardError => e
+    puts "Could not parse metrics: #{e.message}"
+    [0, 0, 0]
   end
 
-  def create_body(metrics, reported_at, status, logs, host)
+  def create_body(metrics, reported_at, status, logs, host, keywords, summary)
     {
-      :migrated => "true",
-      :host => host,
-      :reported_at => reported_at,
-      :logs => logs,
-      :keywords => keywords(status),
-      :summary => summary(metrics),
+      version: 1,
+      migrated: true,
+      host: host.name,
+      reported_at: reported_at,
+      logs: logs,
+      keywords: keywords,
+      summary: summary,
+      # metrics cannot be migrated because Foreman stores them in its own way and
+      # the new host format uses puppet native version
+      metrics: {
+        resources: { values: [] },
+        time: { values: [] },
+        changes: { values: [] },
+        events: { values: [] },
+      },
+      legacy_metrics: metrics,
     }
   end
 
-  def host_report_create(host_id, origin, body)
+  def host_report_create(host_id, origin, body, report_keyword_ids)
     {
-      :host_id => host_id,
-      :proxy_id => nil,
-      :format => origin,
-      :reported_at => body[:reported_at],
-      :body => body.to_json,
-      :change => body[:summary][:foreman][:change],
-      :nochange => body[:summary][:foreman][:nochange],
-      :failure => body[:summary][:foreman][:failure],
+      host_id: host_id,
+      proxy_id: nil,
+      format: origin,
+      reported_at: body[:reported_at],
+      body: body.to_json,
+      change: body[:summary][:foreman][:change],
+      nochange: body[:summary][:foreman][:nochange],
+      failure: body[:summary][:foreman][:failure],
+      report_keyword_ids: report_keyword_ids,
     }
   end
 
+  def create_logs(log_object)
+    logs = []
+    log_object.reorder(:id).find_each do |log|
```

Comment:
We may want to include the source and message here, otherwise `log.source` and `log.message` is loaded separately for each `log_object`

Reply:
Good idea indeed.